### Limpeza da base de dados de notas fiscais


### Descrição das Colunas - Base de Notas Fiscais

| Nome da Coluna           | Descrição                                                                                  | Exemplo                                                                 |
|--------------------------|-------------------------------------------------------------------------------------------|-------------------------------------------------------------------------|
| **Numero nota fiscal**   | Identificador único da nota fiscal (documento fiscal).                                    | `566082`                                                               |
| **Data da venda**        | Data de emissão da nota fiscal (formato DD/MM/AAAA).                                      | `17/06/2025`                                                           |
| **Código produto**       | Código único do produto vendido (deve corresponder ao `Código` na base de estoque).       | `33932`, `241`                                                         |
| **Descrição do produto** | Nome/descrição do produto conforme registrado na nota fiscal.                             | `"ENV 229X324 OU C/100 FILIPERSON"`, `"PASTA AZ OF LE TIGRADA FRAMA"`  |
| **Quantidade do produto**| Quantidade vendida do produto (unidades, caixas, kg, etc.).                              | `2.0`, `40.0`                                                          |
| **Valor unitário**       | Preço unitário do produto no momento da venda (sem descontos/acréscimos).                | `33.000000`, `9.450000`                                                |
| **Preço venda**          | Preço mínimo de venda sugerido.                  | `27.476514`, `8.863195`                                                |
| **Preço de custo**       | Custo unitário do produto no momento da compra (o custo da empresa para adquirir).                  | `24.409041`, `7.605561`                                                |
| **Valor da nota**        | Valor total da nota fiscal (somatório de todos os itens + impostos/fretes, se aplicável).| `3895.94`                                                              |


In [1]:
import pandas as pd

In [2]:
df_vendas = pd.read_excel('bases/relatorio_notas.xlsx')

### Propriedades da base

In [3]:
print(f"Formato dos dados {df_vendas.shape}")

Formato dos dados (336739, 9)


In [34]:
len(df_vendas["Numero nota fiscal"].unique())

30521

In [4]:
df_vendas.columns.tolist()

['Numero nota fiscal',
 'Data da venda',
 'Código produto',
 'Descrição do produto',
 'Quantidade do produto',
 'Valor unitário',
 'Preço venda',
 'Preço de custo',
 'Valor da nota']

In [5]:
df_vendas.head()

Numero nota fiscal Data da venda  Código produto  \
0              566082    17/06/2025           33932   
1              566082    17/06/2025           29367   
2              566082    17/06/2025             241   
3              566082    17/06/2025           32608   
4              566082    17/06/2025           18508   

                       Descrição do produto  Quantidade do produto  \
0           ENV 229X324 OU C/100 FILIPERSON                    2.0   
1     GUARDANAPO SNOB FS 33X30 50F SANTHER                     5.0   
2    PASTA AZ OF LE TIGRADA FRAMA 356-11952                   40.0   
3  CALCULADORA MESA 12 DIG MASTERPRINT 1086                    6.0   
4       CAD BROCH UNIV STIFF AZ 96F JANDAIA                    5.0   

   Valor unitário  Preço venda  Preço de custo  Valor da nota  
0           33.00    27.476514       24.409041        3895.94  
1            2.90     2.499397        2.231242        3895.94  
2            9.45     8.863195        7.605561        3895.94  
3           20.87    18.974913       17.051309        3895.94  
4            6.26     5.680850        4.830296        3895.94

In [6]:
df_vendas.dtypes

Numero nota fiscal         int64
Data da venda             object
Código produto             int64
Descrição do produto      object
Quantidade do produto    float64
Valor unitário           float64
Preço venda              float64
Preço de custo           float64
Valor da nota            float64
dtype: object

In [7]:
df_vendas.describe()

Numero nota fiscal  Código produto  Quantidade do produto  \
count       336739.000000   336739.000000          336739.000000   
mean        480532.632603    22071.594784              15.671684   
std         180938.617261    11782.581172              85.143989   
min            401.000000        5.000000           -3125.000000   
25%         544228.000000    11430.000000               2.000000   
50%         551127.000000    24781.000000               5.000000   
75%         558739.000000    32304.000000              12.000000   
max         566082.000000    38229.000000            9764.000000   

       Valor unitário    Preço venda  Preço de custo  Valor da nota  
count   336739.000000  336739.000000   334515.000000  333994.000000  
mean        15.853359      13.092898       11.769463    2087.380438  
std         28.686929      23.675404       21.230737    3667.545385  
min          0.000000       0.000000        0.000000       0.000000  
25%          3.970000       3.187960        2.833333     502.070000  
50%          8.190000       6.671038        6.007989    1076.750000  
75%         18.820000      15.223790       13.530400    2341.640000  
max       1169.000000     974.762703      880.340302  178016.720000

In [8]:
print("Quantidade valores Nulos")
df_vendas.isnull().sum()

Quantidade valores Nulos


Numero nota fiscal          0
Data da venda               0
Código produto              0
Descrição do produto        0
Quantidade do produto       0
Valor unitário              0
Preço venda                 0
Preço de custo           2224
Valor da nota            2745
dtype: int64

In [9]:
print("Quantidade de valores zerados")
colunas_numericas = df_vendas.select_dtypes(include='number')
for coluna in colunas_numericas:
    at_zeros = (colunas_numericas[coluna] == 0).sum()
    print(f"\n{coluna} contém {at_zeros} zeros")



Quantidade de valores zerados

Numero nota fiscal contém 0 zeros

Código produto contém 0 zeros

Quantidade do produto contém 1749 zeros

Valor unitário contém 2017 zeros

Preço venda contém 1916 zeros

Preço de custo contém 1 zeros

Valor da nota contém 1108 zeros


In [10]:
print("Quantidade de valores negativos")
for rotulo in colunas_numericas:
    qt_negativos = (colunas_numericas[rotulo] < 0).sum()
    print(f"\n{rotulo} contém {qt_negativos} negativos")

Quantidade de valores negativos

Numero nota fiscal contém 0 negativos

Código produto contém 0 negativos

Quantidade do produto contém 459 negativos

Valor unitário contém 0 negativos

Preço venda contém 0 negativos

Preço de custo contém 0 negativos

Valor da nota contém 0 negativos


### Análise de Consistência e Decisões - Base de Vendas

#### **Regras de Negócio Validadas**
| Regra | Descrição | Impacto no Modelo |
|-------|-----------|-------------------|
| **Cálculo do Valor por Item** | `Valor unitário` × `Quantidade produto` = Valor total do produto na nota. | Garante a correta agregação dos valores por item. |
| **Cálculo do Valor da Nota** | Somatório do valor total de todos os itens = `Valor da nota`. | Notas com divergência serão sinalizadas para revisão. |
| **Exclusão de Campo** | `Preço venda` será removido do modelo por ser apenas uma sugestão comercial. | Reduz ruídos e simplifica a análise de preços reais. |

---

#### **Problemas Identificados e Impactos**

1. **Preço de Custo Nulo**  
   - **Ocorrências**: 2.224 registros sem `Preço de custo`.  
   - **Análise**:  
     - Produtos, categorias e marcas afetadas serão priorizadas por volume de vendas.  
     - Impacta o cálculo de margem de lucro para **itens em múltiplas notas fiscais**.  

2. **Notas com Valores Zerados**  
   | Campo Zerado | Ocorrências | Impacto |
   |-------------|------------|---------|
   | `Quantidade produto` | 1.749 | Itens inválidos (exclusão necessária). |
   | `Valor unitário` | 2.017 | Distorção no valor total da nota. |  
   - **Decisão**: Registros com esses campos zerados serão **excluídos** do modelo.  

3. **Valores Negativos**  
   - **Ocorrências**: 459 registros com `Quantidade produto` negativa.  
   - **Ação**: Serão excluidos da base, mas **exportados** para serem analisados separadamente.  

---

#### **Resumo de Ações para o Modelo**
| Etapa | Decisão | Justificativa |
|-------|---------|---------------|
| **Padronização** | Validar `Valor da nota` vs. somatório dos itens. | Garantir integridade financeira. |
| **Limpeza** | Excluir registros com `Quantidade produto` ≤ 0 ou `Valor unitário` = 0. | 0.5% total de notas impactadas Evitar distorções nas recomendações. |
| **Substituição** | Preencher `Preço de custo` nulo com valores da base de estoque. | Permitir cálculo de margem. |
| **Exclusão** | Remover `Preço venda` (campo não confiável). | Foco em dados reais de transação. |

In [11]:
# Filtrar apenas as linhas com quantidade negativa
df_negativos = df_vendas[df_vendas['Quantidade do produto'] < 0]

# Contar produtos únicos e notas fiscais impactadas
total_produtos_negativos = df_negativos['Descrição do produto'].nunique()
total_notas_impactadas = df_negativos['Numero nota fiscal'].nunique()

# Agrupar por produto para ver a distribuição
analise_produtos = df_negativos.groupby('Descrição do produto').agg(
    Total_Quantidade_Negativa=('Quantidade do produto', 'sum'),
    Ocorrencias=('Quantidade do produto', 'count')
).sort_values('Total_Quantidade_Negativa')

print(f"🔍 {len(df_negativos)} registros negativos afetam {total_notas_impactadas} notas fiscais e {total_produtos_negativos} produtos distintos.\n")
print("📊 Produtos com quantidades negativas:")
display(analise_produtos)

🔍 459 registros negativos afetam 49 notas fiscais e 337 produtos distintos.

📊 Produtos com quantidades negativas:


Total_Quantidade_Negativa  \
Descrição do produto                                                  
COPO 150ML TR C/100 FC                                      -3125.0   
COLHER REF BR C/50 PRAFESTA 7024                            -1400.0   
COLHER SOB BR C/50 STRAW CSB620                              -986.0   
VELA ESTRELA PRATA C/4 TOP FESTAS                            -720.0   
PANO CHAO CRU 40X58 L3P2 ALGOBOM                             -608.0   
...                                                             ...   
BL BINGO JORNAL 100F SERIE 16-30 TAMOIO                        -1.0   
BL REC 38X50 AM C/4 100F NOTEFIX 3M NFX3                       -1.0   
SABONETE ERVA DOCE 5L BEM LIMPIM                               -1.0   
TINTA GUACHE AM OU 250ML ACRILEX 505                           -1.0   
ACENDEDOR MULTIUSO HANDY BIC 871354                            -1.0   

                                          Ocorrencias  
Descrição do produto                                   
COPO 150ML TR C/100 FC                              1  
COLHER REF BR C/50 PRAFESTA 7024                    2  
COLHER SOB BR C/50 STRAW CSB620                     2  
VELA ESTRELA PRATA C/4 TOP FESTAS                   2  
PANO CHAO CRU 40X58 L3P2 ALGOBOM                    2  
...                                               ...  
BL BINGO JORNAL 100F SERIE 16-30 TAMOIO             1  
BL REC 38X50 AM C/4 100F NOTEFIX 3M NFX3            1  
SABONETE ERVA DOCE 5L BEM LIMPIM                    1  
TINTA GUACHE AM OU 250ML ACRILEX 505                1  
ACENDEDOR MULTIUSO HANDY BIC 871354                 1  

[337 rows x 2 columns]

In [12]:
# gardando base com valores negativos para uma análise posterior
analise_produtos.to_excel("produtos_com_quantidade_negativa.xlsx")

In [ ]:
# Contagem para 'Quantidade do produto' zerada
qtde_zeros = (df_vendas['Quantidade do produto'] == 0).sum()
produtos_qtde_zero = df_vendas[df_vendas['Quantidade do produto'] == 0]['Quantidade do produto'].nunique()
notas_qtde_zero = df_vendas[df_vendas['Quantidade do produto'] == 0]['Numero nota fiscal'].nunique()

# Contagem para 'Valor unitário' zerado
valor_zeros = (df_vendas['Valor unitário'] == 0).sum()
produtos_valor_zero = df_vendas[df_vendas['Valor unitário'] == 0]['Quantidade do produto'].nunique()
notas_valor_zero = df_vendas[df_vendas['Valor unitário'] == 0]['Numero nota fiscal'].nunique()

# Resultado
print("📊 Análise de Valores Zerados:")
print(f"\n1. Quantidade do produto:")
print(f"- Registros zerados: {qtde_zeros}")
print(f"- Produtos impactados: {produtos_qtde_zero}")
print(f"- Notas fiscais impactadas: {notas_qtde_zero}")

print(f"\n2. Valor unitário:")
print(f"- Registros zerados: {valor_zeros}")
print(f"- Produtos impactados: {produtos_valor_zero}")
print(f"- Notas fiscais impactadas: {notas_valor_zero}")

📊 Análise de Valores Zerados:

1. Quantidade do produto:
- Registros zerados: 1749
- Produtos impactados: 1
- Notas fiscais impactadas: 40

2. Valor unitário:
- Registros zerados: 2017
- Produtos impactados: 29
- Notas fiscais impactadas: 101


In [14]:
# 1. Identificar notas fiscais problemáticas
notas_com_problemas = set()

# Notas com quantidades negativas
notas_com_problemas.update(
    df_vendas[df_vendas['Quantidade do produto'] < 0]['Numero nota fiscal'].unique()
)

# Notas com quantidades zeradas
notas_com_problemas.update(
    df_vendas[df_vendas['Quantidade do produto'] == 0]['Numero nota fiscal'].unique()
)

# Notas com valores unitários zerados
notas_com_problemas.update(
    df_vendas[df_vendas['Valor unitário'] == 0]['Numero nota fiscal'].unique()
)

# 2. Filtrar para manter APENAS notas SEM problemas
df_venda_limpo = df_vendas[~df_vendas['Numero nota fiscal'].isin(notas_com_problemas)].copy()

# 3. Análise de impacto
total_notas_originais = df_vendas['Numero nota fiscal'].nunique()
total_notas_limpas = df_venda_limpo['Numero nota fiscal'].nunique()
notas_removidas = total_notas_originais - total_notas_limpas

print(f"📊 Notas fiscais removidas: {notas_removidas}/{total_notas_originais} ({notas_removidas/total_notas_originais:.1%})")
print(f"📦 Registros originais: {len(df_vendas)}")
print(f"🧹 Registros após limpeza: {len(df_venda_limpo)}")
print(f"🚮 Registros removidos: {len(df_vendas) - len(df_venda_limpo)}")

📊 Notas fiscais removidas: 150/30521 (0.5%)
📦 Registros originais: 336739
🧹 Registros após limpeza: 333804
🚮 Registros removidos: 2935


In [15]:
df_venda_limpo.nunique()

Numero nota fiscal        30371
Data da venda               134
Código produto             9280
Descrição do produto       9105
Quantidade do produto       494
Valor unitário            16560
Preço venda              126645
Preço de custo             9689
Valor da nota             26163
dtype: int64

In [16]:
# 1. Verificar códigos com múltiplas descrições (problema grave)
codigos_problematicos = df_vendas.groupby('Código produto')['Descrição do produto'].nunique()
codigos_com_erro = codigos_problematicos[codigos_problematicos > 1]

print(f"🚨 {len(codigos_com_erro)} códigos com múltiplas descrições:")
display(codigos_com_erro.head())

# 2. Verificar códigos sem descrição (ausência de cadastro)
codigos_sem_descricao = df_vendas[df_vendas['Descrição do produto'].isna()]['Código produto'].nunique()
print(f"\n🔍 {codigos_sem_descricao} códigos sem descrição cadastrada.")

# 3. Análise de correspondência
print("\n📊 Relação Código-Descrição:")
print(f"- Códigos únicos: {df_vendas['Código produto'].nunique()}")
print(f"- Descrições únicas: {df_vendas['Descrição do produto'].nunique()}")
print(f"- Diferença: {df_vendas['Código produto'].nunique() - df_vendas['Descrição do produto'].nunique()}")

🚨 0 códigos com múltiplas descrições:


Series([], Name: Descrição do produto, dtype: int64)


🔍 0 códigos sem descrição cadastrada.

📊 Relação Código-Descrição:
- Códigos únicos: 9434
- Descrições únicas: 9257
- Diferença: 177


In [17]:
# Agrupa descrições e conta quantos códigos únicos cada uma tem
descricoes_problematicas = df_vendas.groupby('Descrição do produto')['Código produto'].nunique()
descricoes_problematicas = descricoes_problematicas[descricoes_problematicas > 1].index.tolist()

print(f"🔍 {len(descricoes_problematicas)} descrições associadas a múltiplos códigos.")

🔍 84 descrições associadas a múltiplos códigos.


In [18]:
# Filtra o DataFrame para manter apenas registros com descrições problemáticas
df_problema = df_vendas[df_vendas['Descrição do produto'].isin(descricoes_problematicas)]

In [19]:
# Conta quantas notas fiscais únicas contêm esses registros
notas_impactadas = df_problema['Numero nota fiscal'].nunique()
total_notas = df_vendas['Numero nota fiscal'].nunique()

print(f"📊 Notas fiscais impactadas: {notas_impactadas}/{total_notas} ({notas_impactadas/total_notas:.1%})")

📊 Notas fiscais impactadas: 3546/30521 (11.6%)


In [20]:
df_problema.to_excel("notas_com_descricoes_ambiguas.xlsx", index=False)

In [21]:
# 1. Encontre descrições associadas a múltiplos códigos
descricoes_ambiguas = df_vendas.groupby('Descrição do produto')['Código produto'].nunique()
descricoes_ambiguas = descricoes_ambiguas[descricoes_ambiguas > 1].index.tolist()

# 2. Filtre as notas fiscais que contêm essas descrições
notas_com_problemas = df_vendas[df_vendas['Descrição do produto'].isin(descricoes_ambiguas)]['Numero nota fiscal'].unique()

In [22]:
# 3. Crie um novo DataFrame EXCLUINDO as notas problemáticas
df_vendas_limpo = df_vendas[~df_vendas['Numero nota fiscal'].isin(notas_com_problemas)].copy()

In [23]:
# 4. Calcule o impacto
total_notas_originais = df_vendas['Numero nota fiscal'].nunique()
total_notas_limpas = df_vendas_limpo['Numero nota fiscal'].nunique()
notas_removidas = total_notas_originais - total_notas_limpas

print(f"🔍 {len(descricoes_ambiguas)} descrições ambíguas encontradas.")
print(f"📊 Notas fiscais removidas: {notas_removidas}/{total_notas_originais} ({notas_removidas/total_notas_originais:.2%})")
print(f"📦 Registros originais: {len(df_vendas)}")
print(f"🧹 Registros após limpeza: {len(df_vendas_limpo)}")

🔍 84 descrições ambíguas encontradas.
📊 Notas fiscais removidas: 3546/30521 (11.62%)
📦 Registros originais: 336739
🧹 Registros após limpeza: 249032


In [24]:
# Tratar o valor da nota, somatório de quantidade de produto * valor unitário
# calcular o valor total de cada produto por nota
df_vendas_limpo["Valor total produto"] = (df_venda_limpo["Quantidade do produto"] * df_venda_limpo["Valor unitário"]).round(2)

In [25]:
# Passo 1: Calcular o valor total por nota fiscal
valor_por_nota = df_vendas_limpo.groupby('Numero nota fiscal')['Valor total produto'].sum().round(2)

# Passo 2: Mapear esse valor de volta para a coluna 'Valor da nota'
df_vendas_limpo['Valor da nota'] = df_vendas_limpo['Numero nota fiscal'].map(valor_por_nota)

# Verificação (opcional)
print("✅ Valores atualizados com sucesso!")
print(f"Total de notas fiscais processadas: {len(valor_por_nota)}")

✅ Valores atualizados com sucesso!
Total de notas fiscais processadas: 26975


In [26]:
df_vendas_limpo = df_vendas_limpo[["Numero nota fiscal", "Data da venda", "Código produto", "Descrição do produto", "Quantidade do produto", "Valor unitário", "Preço de custo", "Valor total produto", "Valor da nota"]]

In [27]:
# Arredonda as colunas específicas para 2 casas decimais
df_vendas_limpo['Valor unitário'] = df_vendas_limpo['Valor unitário'].round(2)
df_vendas_limpo['Preço de custo'] = df_vendas_limpo['Preço de custo'].round(2)

In [28]:
df_vendas_limpo.to_excel("bases/bases_limpas/base_notas_limpa.xlsx", index=False)

In [37]:
df_vendas_limpo.nunique()

Numero nota fiscal       26975
Data da venda              135
Código produto            8735
Descrição do produto      8735
Quantidade do produto      506
Valor unitário            9028
Preço de custo            3621
Valor total produto      23828
Valor da nota            22506
dtype: int64